src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java

```diff
@@ -0,0 +1,364 @@
+	{
+		doFtpDownload(null, null);
+	}
+	
+	protected void doFtpDownload(String user, String password) throws SocketException, IOException, FileNotFoundException, Exception
+	{
+		user = user == null || user.trim().equals("") ? "anonymous" : user;
+		password = password == null || password.trim().equals("") ? "" : password;
+		FTPClient client = new FTPClient();
+		
+		client.connect(this.uri.getHost());
+		if (this.passiveFTP)
+		{
+			client.enterLocalPassiveMode(); //PASSIVE mode works better when inside a docker container.
+		}
+		client.login(user, password);
+		logger.debug("connect/login reply code: {}",client.getReplyCode());
+		client.setFileType(FTP.BINARY_FILE_TYPE);
+		client.setFileTransferMode(FTP.COMPRESSED_TRANSFER_MODE);
+		try
+		{
+			try(InputStream inStream = client.retrieveFileStream(this.uri.getPath()))
+			{
+				if (inStream != null)
+				{
+					writeInputStreamToFile(inStream);
+				}
+				else
+				{
+					logger.error("No data returned from server for {}", this.uri.toString());
+				}
+			}
+		}
+		catch (IOException e)
+		{
+			logger.error("Error while retrieving the file: {}",e.getMessage());
+			e.printStackTrace();
+			throw new IOException(e);
+		}
+		//Should probably have more/better reply-code checks.
+		logger.debug("retreive file reply code: {}",client.getReplyCode());
+		if (client.getReplyString().matches("^5\\d\\d.*") || (client.getReplyCode() >= 500 && client.getReplyCode() < 600) )
+		{
+			String errorString = "5xx reply code detected (" + client.getReplyCode() + "), reply string is: "+client.getReplyString();
+			logger.error(errorString);
+			throw new Exception(errorString);
+		}
+		client.logout();
+		client.disconnect();
+	}
+
+	protected void writeInputStreamToFile(InputStream inStream) throws IOException, FileNotFoundException
+	{
+		try(ByteArrayOutputStream baos = new ByteArrayOutputStream())
+		{
+			int b = inStream.read();
+			while (b!=-1)
+			{
+				baos.write(b);
+				b = inStream.read();
+			}
+	
+			try (FileOutputStream file = new FileOutputStream(this.destination))
+			{
+				baos.writeTo(file);
+				file.flush();
+			}
+		}
+	}
+
+	
+	protected void doHttpDownload(Path path) throws HttpHostConnectException, IOException, Exception
+	{
+		this.doHttpDownload(path, HttpClientContext.create());
+	}
+	
+	protected void doHttpDownload(Path path, HttpClientContext context) throws Exception, HttpHostConnectException, IOException
+	{
+		HttpGet get = new HttpGet(this.uri);
+		//Need to multiply by 1000 because timeouts are in milliseconds.
+		RequestConfig config = RequestConfig.copy(RequestConfig.DEFAULT)
+											.setConnectTimeout(1000 * (int)this.timeout.getSeconds())
+											.setSocketTimeout(1000 * (int)this.timeout.getSeconds())
+											.setConnectionRequestTimeout(1000 * (int)this.timeout.getSeconds()).build();
+		
+		get.setConfig(config);
+		
+		int retries = this.numRetries;
+		boolean done = retries + 1 <= 0;
+		while(!done)
```

Comment:
Is it the case this block is always expected to run once (i.e. the number of retries should always be 0 or more)?  If so, it might be better to use a "do-while" loop so the condition is evaluated at the end and `boolean done` is declared outside the loop but only assigned inside.

Reply:
I like while-loops because you can read the condition at the top. I find it preferable to looking for the end of the loop to understand the loop condition.

Reply:
I think it might be warranted to a do loop in this case (or a while loop with an internal break) because the loop condition is changed based on a successful download or running out of retries which requires looking in the loop body to understand since the condition is stored in the `done` variable which is set outside and altered inside rather than described in the while statement at the top itself.

In any case, I find it confusing that retries becomes -1 to end the loop.  Intuitively, I would expect that retries would be 0 or more.  How the loop is structured is up to you, but that condition really makes confused with the code and would be better if it could exit the loop when it finds the download failed and the retries are 0.

Reply:
I may consider making some changes...

Reply:
I think rewriting this code is out of scope for this PR - it is old, existing code (from AddLinks) that works fine. Changes may be made in the future but I think the effort to change and test would go beyond the scope for this PR which was to integrate the code.